### ISSUE TITLE
Quick security review

### ISSUE THREAD

**Issue by WildCryptoFox** [AndrewBelt/hack.chat#122]
After a **very** quick analysis over the code, I have deducted the following issues.

## Problems

- The administrator password is stored in plaintext and compared in non-constant time. This introduces a timing side-channel that enables an adversary to guess each character of a password recursively until success.

- Password hashing uses `sha256(salt + password)` instead of `HMAC(salt, password)`, use HMAC for any keyed-hashes and **NEVER TRUNCATE HASHES**!

- Truncated hashes (trip codes) may be trivially brute forced (modulo easy to evade rate limitations). As the hashes are truncated, the chances of collisions are significantly increased.

## Quick solutions

- Hash the administrator and moderator passwords as any other user.

- Use `HMAC(salt, password)` or ideally `HMAC(HMAC(salt, username), password)` for "better" password hashing. **NEVER TRUNCATE HASHES**!

- User password hashes should never (even truncated) be shared with other users. It would be better to register users and display the username instead. IRC does exactly this.

## Better solution (pick one)

- (Requires accounts) Use a [Password-authenticated Key Agreement](https://en.wikipedia.org/wiki/Password-authenticated_key_agreement) solution instead so the server never needs to learn passwords.

- (Does not require accounts) Authentication may be delegated; handled by clients/users instead. I advocate using 3DH + Double Ratchet (like Signal). Service bots may rely on this for authenticated transactions (ban/unban, kick, ..).

## TL;DR

The administrator's password is most vulnerable (timing channel) and all users and moderators may have their tripcodes targeted (brute force) because the tripcodes are truncated, significantly increasing the chances of collisions. In the case of moderators, the adversary gains the moderator privileges.

Alternatively all forms of authentication may be removed, we're after minimalism right?

Happy hacking.

**Comment by AndrewBelt** [comment]
Although appreciated, this issue is mostly a duplicate of #119. I'm not really interested in theoretical cryptographic vulnerabilities but definitely interested in practical ones, such as #107. For example, the first solution you propose involves very real changes in UI to solve a currently non-real problem. I'm closing this for now, unless you can tell me the administrator password of https://hack.chat or even the first character.

**Comment by WildCryptoFox** [comment]
No change to the UI is required.

> Use HMAC(salt, password) or ideally HMAC(HMAC(salt, username), password) for "better" password hashing. NEVER TRUNCATE HASHES!

To be clear; using `HMAC(salt, password)` is my primary recommendation; no UI change.

Additionally the tripcode has minimal benefit at all. It allows users to distinguish user A and user B with the same nickname (I.e. impersonation attempts). Because of this the tripcodes do not need to be constant for a user. They may instead be unique between a pair of users.

Without my recommended changes any user (or moderator) may be impersonated by colliding the tripcode.

I have no intention in weaponizing the timing attack to demonstrate the administrator password recovery. If you, or any user of your software uses a password that matters to them. An adversary may recover it.

My recommendation for the administrator and moderator authentication remains: just hash it.

## TL;DR

- Use `HMAC(salt, password)` not `HASH(salt + password)`.
- Hash administrator and moderator passwords, do not truncate, compare hashes in constant time.
- Either forget all about tripcodes or implement them without revealing a trivial to brute force identifier.

I will happily create a patch for the first two recommendations; the last is for you to decide.

Your call.

**Comment by AndrewBelt** [comment]
Changing the way tripcodes are generated would break the secure identities of existing users, so people would have trouble verifying themselves after a change like this.

**Comment by WildCryptoFox** [comment]
@AndrewBelt Sure, it is a breaking change. So let it be optional and let the administrator decide if/when to upgrade.

- User tripcodes will be updated

- Administrator and moderator **hashes** must be updated in the configuration file.
